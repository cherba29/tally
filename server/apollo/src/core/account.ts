import { Month } from './month';

export enum Type {
  BILL = 'bill',
  CHECKING = 'checking',
  CREDIT = 'credit',
  CREDIT_CARD = 'credit-card',
  DEFERRED_INCOME = 'deferred income',
  EXTERNAL = 'external',
  INCOME = 'income',
  INVESTMENT = 'investment',
  RETIREMENT = 'retirement',
  SUMMARY = '_summary_',
  TAX = 'tax_',
}

export interface InitData {
  name: string;
  description?: string;
  type: Type;
  number?: string;
  openedOn?: Month;
  closedOn?: Month;
  owners: string[];
}

export class Account {
  readonly name: string;
  readonly description?: string;
  readonly type: Type;
  readonly number?: string;
  readonly openedOn?: Month;
  readonly closedOn?: Month;
  readonly owners: string[];

  constructor(data: InitData) {
    this.name = data.name;
    this.description = data.description;
    this.type = data.type;
    this.number = data.number;
    this.openedOn = data.openedOn;
    this.closedOn = data.closedOn;
    this.owners = data.owners;
  }

  isClosed(month: Month) : boolean {
    return this.closedOn?.isLess(month)  // After closed.
        || !this.openedOn?.isLess(month);  // Before open.
  }
}