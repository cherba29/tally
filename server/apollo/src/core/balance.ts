export enum Type {
  UNKNOWN,
  CONFIRMED,
  PROJECTED,
  AUTO_PROJECTED
}

const combineTypes = (t1: Type, t2: Type): Type => (t1 < t2 ? t2 : t1);

export class Balance {
  constructor(readonly amount: number, readonly date: Date, readonly type: Type) {}

  static negated(balance: Balance): Balance {
    return new Balance(-balance.amount, balance.date, balance.type);
  }

  static add(balance1: Balance, balance2: Balance): Balance {
    const maxDate = balance1.date < balance2.date ? balance2.date : balance1.date;
    return new Balance(
      balance1.amount + balance2.amount,
      maxDate,
      combineTypes(balance1.type, balance2.type)
    );
  }

  static subtract(balance1: Balance, balance2: Balance): Balance {
    return Balance.add(balance1, new Balance(-balance2.amount, balance2.date, balance2.type));
  }

  toString(): string {
    return `Balance { amount: ${this.amount}, date: ${this.date
      .toISOString()
      .slice(0, 10)}, type: ${Type[this.type]} }`;
  }
}
