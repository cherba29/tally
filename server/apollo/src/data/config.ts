import { readFileSync } from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';

interface BudgetPeriod {
  start: string;
  end: string;
}

interface TallyConfig {
  budget_period: BudgetPeriod;
}

export function loadTallyConfig(): TallyConfig {
  if (!process.env.TALLY_FILES) {
    throw new Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  const configPath = path.join(process.env.TALLY_FILES, '_config.yaml');
  const config = yaml.safeLoad(readFileSync(configPath, 'utf-8'));
  if (!config || typeof config === 'string') {
    throw new Error(`Could not parse file at "${configPath}"`);
  }
  if (!('start' in config)) {
    throw new Error(`File "${configPath}" does not specify start`);
  }
  if (!('end' in config)) {
    throw new Error(`File "${configPath}" does not specify end`);
  }
  return config as TallyConfig;
}
