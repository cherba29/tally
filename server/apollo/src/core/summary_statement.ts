import {Balance} from './balance';
import {Transaction} from './transaction';


export interface SummaryStatement {
  isClosed: boolean;
  accounts: string[];
  addSub: number;
  endBalance: Balance;
  isCovered: boolean;
  isProjectedCovered: boolean;
  hasProjectedTransfer: boolean;
  percentChange: number;
  unaccounted: number;
  transactions: Transaction[];
}
