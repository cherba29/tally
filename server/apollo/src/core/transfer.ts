import { Account } from "./account";
import { Balance } from "./balance";
import { Month } from "./month";

export interface Transfer {
  fromAccount: Account;
  toAccount: Account;
  fromMonth: Month;
  toMonth: Month;
  description?: string;
  balance: Balance;
}