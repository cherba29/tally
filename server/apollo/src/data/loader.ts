import * as fs from 'fs';
import * as path from 'path';

import { Budget, BudgetBuilder } from '@tally/lib';
import { loadYamlFile } from './loader_yaml';
import { loadTallyConfig } from './config';
// import {SummaryStatement} from '../core/summary_statement';

function* readdirSync(file_path: string): Generator<string> {
  if (fs.statSync(file_path).isDirectory()) {
    for (const f of fs.readdirSync(file_path)) {
      const child_file = path.join(file_path, f);
      for (const cf of readdirSync(child_file)) {
        if (cf.match(/\.yaml$/)) {
          yield cf;
        }
      }
    }
  } else {
    yield file_path;
  }
  return;
}

export function listFiles(): string[] {
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  const filePaths: string[] = [];
  for (const file_path of readdirSync(process.env.TALLY_FILES)) {
    const relative_file_path = file_path.slice(process.env.TALLY_FILES.length + 1);
    filePaths.push(relative_file_path);
  }
  return filePaths;
}

export function loadBudget(): Budget {
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  const budgetBuilder = new BudgetBuilder();

  const config = loadTallyConfig();
  // Always show one more month, in case user chooses in the config
  // just single month. It is not meaningful to show only one month
  // as values from next used for current month.
  budgetBuilder.setPeriod(config.budget_period.start, config.budget_period.end.next());

  const filePaths: string[] = [];
  for (const file_path of readdirSync(process.env.TALLY_FILES)) {
    const relative_file_path = file_path.slice(process.env.TALLY_FILES.length + 1);
    filePaths.push(relative_file_path);
    console.log('Loading ' + relative_file_path);
    const content = fs.readFileSync(file_path, 'utf8');
    loadYamlFile(budgetBuilder, content, relative_file_path);
  }
  console.log(`Done loading ${filePaths.length} file(s)`);
  return budgetBuilder.build();
}
