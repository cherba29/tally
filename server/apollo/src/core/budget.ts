import {Account} from './account';
import {Balance} from './balance';
import {Month} from './month';
import {Transfer} from './transfer';

export class Budget {
  // Period over which the budget is defined.
  months: Month[] = [];
  // Account name to account map.
  readonly accounts = new Map<string, Account>();
  // Account name -> month -> balance map.
  readonly balances = new Map<string, Map<string, Balance>>();
  // Account name -> month -> transfers map.
  readonly transfers = new Map<string, Map<string, Set<Transfer>>>();

  setPeriod(start: Month, end: Month) {
    this.months = Array.from(Month.generate(start, end));
  }

  setAccount(account: Account) {
    this.accounts.set(account.name, account);
  }

  setBalance(accountName: string, month: string, balance: Balance) {
    let balances = this.balances.get(accountName);
    if (!balances) {
      balances = new Map<string, Balance>();
      this.balances.set(accountName, balances);
    }
    balances.set(month, balance);
  }
}