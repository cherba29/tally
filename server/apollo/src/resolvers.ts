import {Month} from './core/month';
import { loadTallyConfig } from './data/config'
import { listFiles, loadStatements, loadAccounts } from './data/loader'
import { GraphQLScalarType, Kind } from 'graphql';

export default {
  Query: {
    accounts: loadAccounts,
    files: listFiles,
    statements: loadStatements,
    months: (): Month[] => {
      const config = loadTallyConfig();
      const start = Month.fromString(config.budget_period.start);
      const end = Month.fromString(config.budget_period.end).next();
      return Array.from(Month.generate(start, end));
    },
  },
  Date: new GraphQLScalarType({
    name: 'Date',
    description: 'Date representation in YYYY-MM-DD format.',
    parseValue(value: string): Date {
      return new Date(value);  // value from the client
    },
    serialize(value: Date): string {
      return value.toISOString().split('T')[0];
    },
    parseLiteral(ast): Date | null {
      if (ast.kind === Kind.STRING) {
        return new Date(ast.value);
      }
      return null;
    },
  }),
  GqlMonth: new GraphQLScalarType({
    name: 'GqlMonth',
    description: 'Month representation in XxxYYYY format.',
    parseValue(value: string): Month {
      return Month.fromString(value);  // value from the client
    },
    serialize(value: Month): string {
      return value.toString();
    },
    parseLiteral(ast): Month | null {
      if (ast.kind === Kind.STRING) {
        return Month.fromString(ast.value);
      }
      return null;
    },
  }),
};
