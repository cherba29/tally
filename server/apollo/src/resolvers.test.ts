import { Month } from '@tally/lib/core/month';
import { unwatchBudgetFiles } from '@tally/lib/data/loader';
import resolvers from './resolvers';

describe('date', () => {
  test('parseValue', () => {
    expect(resolvers.Date.parseValue('2021-03-05')).toEqual(new Date('2021-03-05'));
  });
});

describe('GqlMonth', () => {
  test('parseValue', () => {
    expect(resolvers.GqlMonth.parseValue('Mar2021')).toEqual(new Month(2021, 2));
  });
});

describe('gqlTable', () => {
  beforeEach(() => {
    process.env.TALLY_FILES = 'server/apollo/src/testdata/tally';
  });

  test('load', async () => {
    const data = await resolvers.Query.table(undefined, { owner: 'john' });
    expect(data).toMatchSnapshot();
  });

  afterEach(() => {
    unwatchBudgetFiles();
    delete process.env.TALLY_FILES;
  });
});

describe('gqlSummary', () => {
  beforeEach(() => {
    process.env.TALLY_FILES = 'server/apollo/src/testdata/tally';
  });

  test('load', async () => {
    const data = await resolvers.Query.summary(undefined, {
      owner: 'john',
      month: 'Feb2015',
      accountType: 'BILL'
    });
    expect(data).toMatchSnapshot();
  });

  afterEach(() => {
    unwatchBudgetFiles();
    delete process.env.TALLY_FILES;
  });
});
