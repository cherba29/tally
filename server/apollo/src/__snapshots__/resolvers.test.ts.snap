// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gqlSummary load 1`] = `
{
  "statements": [
    {
      "addSub": 0,
      "annualizedPercentChange": 60.51,
      "change": -58,
      "endBalance": {
        "amount": -1500,
        "date": 2015-03-02T00:00:00.000Z,
        "type": "PROJECTED",
      },
      "hasProjectedTransfer": false,
      "inFlows": 0,
      "income": 0,
      "isClosed": false,
      "isCovered": false,
      "isProjectedCovered": false,
      "month": Month {
        "month": 1,
        "year": 2015,
      },
      "name": "gas",
      "outFlows": 0,
      "percentChange": 4.02,
      "startBalance": {
        "amount": -1442,
        "date": 2015-02-02T00:00:00.000Z,
        "type": "CONFIRMED",
      },
      "totalPayments": 0,
      "totalTransfers": 0,
      "transactions": [],
      "unaccounted": -58,
    },
  ],
  "total": {
    "accounts": [
      "gas",
    ],
    "addSub": 0,
    "annualizedPercentChange": 60.51,
    "change": -58,
    "endBalance": {
      "amount": -1500,
      "date": 2015-03-02T00:00:00.000Z,
      "type": "PROJECTED",
    },
    "inFlows": 0,
    "income": 0,
    "month": Month {
      "month": 1,
      "year": 2015,
    },
    "name": "john BILL",
    "outFlows": 0,
    "percentChange": 4.02,
    "startBalance": {
      "amount": -1442,
      "date": 2015-02-02T00:00:00.000Z,
      "type": "CONFIRMED",
    },
    "totalPayments": 0,
    "totalTransfers": 0,
    "unaccounted": -58,
  },
}
`;

exports[`gqlTable load 1`] = `
{
  "currentOwner": "john",
  "months": [
    Month {
      "month": 3,
      "year": 2015,
    },
    Month {
      "month": 2,
      "year": 2015,
    },
    Month {
      "month": 1,
      "year": 2015,
    },
    Month {
      "month": 0,
      "year": 2015,
    },
    Month {
      "month": 11,
      "year": 2014,
    },
  ],
  "owners": [
    "john",
    "xxx",
  ],
  "rows": [
    {
      "cells": [
        {
          "addSub": 0,
          "balance": undefined,
          "balanced": true,
          "isProjected": true,
          "month": Month {
            "month": 3,
            "year": 2015,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
        {
          "addSub": 0,
          "balance": -1500,
          "balanced": true,
          "isProjected": true,
          "month": Month {
            "month": 2,
            "year": 2015,
          },
          "percentChange": 0,
          "unaccounted": 0,
        },
        {
          "addSub": 0,
          "balance": -1500,
          "balanced": false,
          "isProjected": true,
          "month": Month {
            "month": 1,
            "year": 2015,
          },
          "percentChange": 4.02,
          "unaccounted": -58,
        },
        {
          "addSub": -4912,
          "balance": -1442,
          "balanced": false,
          "isProjected": false,
          "month": Month {
            "month": 0,
            "year": 2015,
          },
          "percentChange": 283.51,
          "unaccounted": 3846,
        },
        {
          "addSub": 0,
          "balance": -376,
          "balanced": true,
          "isProjected": false,
          "month": Month {
            "month": 11,
            "year": 2014,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
      ],
      "isTotal": true,
      "title": "john",
    },
    {
      "cells": [],
      "isSpace": true,
      "title": "BILL accounts",
    },
    {
      "account": {
        "address": undefined,
        "external": false,
        "name": "gas",
        "number": "1234567",
        "openedOn": Month {
          "month": 8,
          "year": 2014,
        },
        "owners": [
          "john",
        ],
        "password": "password",
        "path": [],
        "phone": undefined,
        "summary": false,
        "type": "BILL",
        "url": "http://www.gas.com/",
        "userName": "john",
      },
      "cells": [
        {
          "addSub": 0,
          "balance": undefined,
          "balanced": true,
          "hasProjectedTransfer": false,
          "isClosed": false,
          "isCovered": true,
          "isProjected": false,
          "isProjectedCovered": true,
          "month": Month {
            "month": 3,
            "year": 2015,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
        {
          "addSub": 0,
          "balance": -1500,
          "balanced": true,
          "hasProjectedTransfer": false,
          "isClosed": false,
          "isCovered": false,
          "isProjected": true,
          "isProjectedCovered": false,
          "month": Month {
            "month": 2,
            "year": 2015,
          },
          "percentChange": 0,
          "unaccounted": 0,
        },
        {
          "addSub": 0,
          "balance": -1500,
          "balanced": false,
          "hasProjectedTransfer": false,
          "isClosed": false,
          "isCovered": false,
          "isProjected": true,
          "isProjectedCovered": false,
          "month": Month {
            "month": 1,
            "year": 2015,
          },
          "percentChange": 4.02,
          "unaccounted": -58,
        },
        {
          "addSub": -847,
          "balance": -1442,
          "balanced": false,
          "hasProjectedTransfer": false,
          "isClosed": false,
          "isCovered": false,
          "isProjected": false,
          "isProjectedCovered": false,
          "month": Month {
            "month": 0,
            "year": 2015,
          },
          "percentChange": -8.15,
          "unaccounted": 975,
        },
        {
          "addSub": 0,
          "balance": -1570,
          "balanced": true,
          "hasProjectedTransfer": false,
          "isClosed": false,
          "isCovered": true,
          "isProjected": false,
          "isProjectedCovered": true,
          "month": Month {
            "month": 11,
            "year": 2014,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
      ],
      "isNormal": true,
      "title": "gas",
    },
    {
      "account": {
        "address": undefined,
        "closedOn": Month {
          "month": 0,
          "year": 2015,
        },
        "external": false,
        "name": "phone",
        "number": "1234567890",
        "openedOn": Month {
          "month": 0,
          "year": 2011,
        },
        "owners": [
          "john",
        ],
        "password": "password",
        "path": [],
        "phone": undefined,
        "summary": false,
        "type": "BILL",
        "url": "http://verizon.com",
        "userName": "john",
      },
      "cells": [
        {
          "addSub": 0,
          "balance": undefined,
          "balanced": true,
          "hasProjectedTransfer": false,
          "isClosed": true,
          "isCovered": true,
          "isProjected": false,
          "isProjectedCovered": true,
          "month": Month {
            "month": 3,
            "year": 2015,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
        {
          "addSub": 0,
          "balance": undefined,
          "balanced": true,
          "hasProjectedTransfer": false,
          "isClosed": true,
          "isCovered": true,
          "isProjected": false,
          "isProjectedCovered": true,
          "month": Month {
            "month": 2,
            "year": 2015,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
        {
          "addSub": 0,
          "balance": undefined,
          "balanced": true,
          "hasProjectedTransfer": false,
          "isClosed": true,
          "isCovered": true,
          "isProjected": false,
          "isProjectedCovered": true,
          "month": Month {
            "month": 1,
            "year": 2015,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
        {
          "addSub": -4065,
          "balance": 0,
          "balanced": false,
          "hasProjectedTransfer": false,
          "isClosed": false,
          "isCovered": true,
          "isProjected": false,
          "isProjectedCovered": true,
          "month": Month {
            "month": 0,
            "year": 2015,
          },
          "percentChange": -100,
          "unaccounted": 2871,
        },
        {
          "addSub": 0,
          "balance": 1194,
          "balanced": true,
          "hasProjectedTransfer": false,
          "isClosed": false,
          "isCovered": true,
          "isProjected": false,
          "isProjectedCovered": true,
          "month": Month {
            "month": 11,
            "year": 2014,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
      ],
      "isNormal": true,
      "title": "phone",
    },
    {
      "cells": [
        {
          "addSub": 0,
          "annualizedPercentChange": undefined,
          "balance": undefined,
          "balanced": true,
          "isClosed": false,
          "isProjected": true,
          "month": Month {
            "month": 3,
            "year": 2015,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
        {
          "addSub": 0,
          "annualizedPercentChange": 0,
          "balance": -1500,
          "balanced": true,
          "isClosed": false,
          "isProjected": true,
          "month": Month {
            "month": 2,
            "year": 2015,
          },
          "percentChange": 0,
          "unaccounted": 0,
        },
        {
          "addSub": 0,
          "annualizedPercentChange": 60.51,
          "balance": -1500,
          "balanced": false,
          "isClosed": false,
          "isProjected": true,
          "month": Month {
            "month": 1,
            "year": 2015,
          },
          "percentChange": 4.02,
          "unaccounted": -58,
        },
        {
          "addSub": -4912,
          "annualizedPercentChange": undefined,
          "balance": -1442,
          "balanced": false,
          "isClosed": false,
          "isProjected": false,
          "month": Month {
            "month": 0,
            "year": 2015,
          },
          "percentChange": 283.51,
          "unaccounted": 3846,
        },
        {
          "addSub": 0,
          "annualizedPercentChange": undefined,
          "balance": -376,
          "balanced": true,
          "isClosed": false,
          "isProjected": false,
          "month": Month {
            "month": 11,
            "year": 2014,
          },
          "percentChange": undefined,
          "unaccounted": undefined,
        },
      ],
      "isTotal": true,
      "title": "BILL",
    },
  ],
}
`;
